### ISSUE TITLE
training: batched input (B>1)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#7]
**Independently valuable** training-maturity work + the actual pre-requisite for #5 (DP training is moot without batch>1).

## What's needed

Today's training path processes **one sequence per step** (batch=1). The bottleneck:

- `LlamaSeqForwardFFICache#realize_for_random_init(cfg, T, …)` allocates `[T, 1]` shaped inputs
- `realize_for_mmap`, `realize_for_full_finetune` mirror this
- All three backends (CPU / CUDA / Metal) follow the same shape

To enable batch>1:

- `realize_for_*(cfg, T, B, …)` — accept a batch dim
- `t_seq_token_ids` becomes `[T, B]` (i32)
- attention mask + RoPE generalise (mostly ggml-internal; the wiring change is in realize + upload sites)
- labels tensor becomes `[T×B, vocab]` (same one-hot pattern)
- loss reduction extends from mean-over-T to mean-over-T×B (already does the T-side)

## Scope

~100-200 LOC across `lib/llama_seq_forward_ffi{,_cuda,_metal}.rb`. CPU mirror in tinynn lib stays unchanged (the math is in ggml).

## Why this is independently valuable

Even single-GPU throughput goes up with batch>1 — matmul utilization grows linearly with batch up to the kernel's saturation point. So this is a "real training speedup" toy can ship *before* any multi-GPU work.

## Acceptance

```sh
TAO_RUN_DIR=/tmp/r BATCH=4 STEPS=20 ./examples/example_train_from_scratch
```

trains with batch=4, emits `run_start.config.batch == 4`, loss decreases monotonically, `step.tokens == B×T` per step.

## References

- `docs/roadmap/backends-and-scale-2026-05-27.md` §"Batched training input" for the LOC breakdown.
- Issue #5 lists batching as the gating dependency for DP training.

**Comment by OriPekelman** [comment]
**Shipped** in 87800fa (toy#7: micro-batching).

`realize_for_random_init` gains a `t_batch` arg (now 7 positionals). B=1 stays bit-identical to pre-GH#7; B>1 lays B sequences side-by-side as a flat `[T*B]` token + position vector with a block-causal attention mask uploaded at realize.

**Math**
- Token IDs + positions length `T*B`; positions cycle `0..T-1` per batch element so RoPE applies per-batch positional encoding.
- Attention mask: `f32 [T*B, T*B]` persistent in ctx_w. `0.0` for (query, key) pairs in the same batch with `key <= query`; `-1.0e30` otherwise (`exp(-1e30) == 0` in f32). Applied via `tnn_soft_max_ext(scores, mask, scale, 0.0)` — folds scale + mask + softmax into one op.
- B=1 keeps the legacy `scale + diag_mask_inf + softmax` triple unchanged; the conditional is `if @seq_b > 1` in `build_seq_qhead`. Everywhere else `T*B` arithmetic collapses to `T` at B=1.

**Verification**
- CPU B=1 step-1 CE = 6.490187 (matches pre-GH#7 to f32).
- CUDA B=1 step-1 CE = 6.490198 (matches CPU within float noise).
- CPU + CUDA both learn at B=4, B=8.
- `T=64 B=8 STEPS=20` CUDA: CE 6.51 → 4.84 in 12 ms/step.
- CUDA's per-step time drops with B=8 (19.5 ms vs 24.2 ms at B=1, T=16) — launch overhead amortizes.
- LoRA finetune (`realize_for_mmap`, separate codepath) still trains — unaffected.

**Usage** (`examples/06_train_from_scratch.rb`):
```
BATCH=4 CONTEXT=32 STEPS=20 ./examples/example_train_from_scratch
```

Closes.
